cpp/src/io/utilities/type_inference.cuh

```diff
@@ -0,0 +1,303 @@
+      continue;
+    }
+
+    uint32_t digit_count    = 0;
+    uint32_t decimal_count  = 0;
+    uint32_t slash_count    = 0;
+    uint32_t dash_count     = 0;
+    uint32_t plus_count     = 0;
+    uint32_t colon_count    = 0;
+    uint32_t exponent_count = 0;
+    uint32_t other_count    = 0;
+
+    auto const maybe_hex = (field_len > 2 && *field_begin == '0' && *(field_begin + 1) == 'x') ||
+                           (field_len > 3 && *field_begin == '-' && *(field_begin + 1) == '0' &&
+                            *(field_begin + 2) == 'x');
+    auto const field_end = field_begin + field_len;
+
+    for (auto pos = field_begin; pos < field_end; ++pos) {
+      if (is_digit(*pos, maybe_hex)) {
+        digit_count++;
+        continue;
+      }
+      // Looking for unique characters that will help identify column types
+      switch (*pos) {
+        case '.': decimal_count++; break;
+        case '-': dash_count++; break;
+        case '+': plus_count++; break;
+        case '/': slash_count++; break;
+        case ':': colon_count++; break;
+        case 'e':
+        case 'E':
+          if (!maybe_hex && pos > field_begin && pos < field_end - 1) exponent_count++;
+          break;
+        default: other_count++; break;
+      }
+    }
+
+    // Integers have to have the length of the string
+    auto int_req_number_cnt = static_cast<uint32_t>(field_len);
+    // Off by one if they start with a minus sign
+    if ((*field_begin == '-' || *field_begin == '+') && field_len > 1) { --int_req_number_cnt; }
+    // Off by one if they are a hexadecimal number
+    if (maybe_hex) { --int_req_number_cnt; }
+    if (cudf::detail::serialized_trie_contains(
+          options.trie_true, {field_begin, static_cast<std::size_t>(field_len)}) ||
+        cudf::detail::serialized_trie_contains(
+          options.trie_false, {field_begin, static_cast<std::size_t>(field_len)})) {
+      atomicAdd(&column_info->bool_count, 1);
+    } else if (digit_count == int_req_number_cnt) {
+      bool is_negative       = (*field_begin == '-');
+      char const* data_begin = field_begin + (is_negative || (*field_begin == '+'));
+      cudf::size_type* ptr   = cudf::io::gpu::infer_integral_field_counter(
+        data_begin, data_begin + digit_count, is_negative, *column_info);
+      atomicAdd(ptr, 1);
+    } else if (is_like_float(
+                 field_len, digit_count, decimal_count, dash_count + plus_count, exponent_count)) {
+      atomicAdd(&column_info->float_count, 1);
+    }
+    // A date field can have either one or two '-' or '\'; A legal combination will only have one
+    // of them To simplify the process of auto column detection, we are not covering all the
+    // date-time formation permutations
+    else if (((dash_count > 0 && dash_count <= 2 && slash_count == 0) ||
+              (dash_count == 0 && slash_count > 0 && slash_count <= 2)) &&
+             colon_count <= 2) {
+      atomicAdd(&column_info->datetime_count, 1);
+    }
+  }  // for
```

Comment:
Just curious if the `// for` was intentional?

Reply:
yeah I've updated the comment to make it more descriptive: `// grid-stride for loop`